File: src/main/java/ru/arc/common/ServerLocation.java
package ru.arc.common;

import ru.arc.ARC;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;

import java.util.Optional;

@AllArgsConstructor
@Builder
@Data
@NoArgsConstructor
public class ServerLocation {

    String server, world;
    double x,y,z;
    float yaw,pitch;

    public static ServerLocation of(Location loc) {
        return new ServerLocation(ARC.serverName, loc.getWorld().getName(),
                loc.getX(), loc.getY(), loc.getZ(), loc.getYaw(), loc.getPitch());
    }

    public Location toLocation(){
        World world1 = Bukkit.getWorld(this.world);
        if(world1 == null) return null;
        return new Location(world1, x, y, z);
    }

    public Optional<Double> distance(Location location){
        if(!ARC.serverName.equals(server)) return Optional.empty();
        if(!world.equals(location.getWorld().getName())) return Optional.empty();
        double distance = Math.pow(location.x() - x, 2) + Math.pow(location.y() - y, 2) + Math.pow(location.z() - z, 2);
        return Optional.of(Math.sqrt(distance));
    }

    public boolean isSameServer(){
        return server.equals(ARC.serverName);
    }

}

File: src/main/kotlin/ru/arc/common/ServerLocation.kt
package ru.arc.common

import org.bukkit.Bukkit
import org.bukkit.Location
import ru.arc.ARC
import kotlin.math.sqrt

data class ServerLocation(
    val server: String? = null,
    val world: String? = null,
    val x: Double = 0.0,
    val y: Double = 0.0,
    val z: Double = 0.0,
    val yaw: Float = 0.0f,
    val pitch: Float = 0.0f,
) {
    companion object {
        @JvmStatic
        fun of(loc: Location?): ServerLocation {
            requireNotNull(loc) { "Location cannot be null" }
            val world = requireNotNull(loc.world) { "Location world cannot be null" }
            return ServerLocation(
                server = ARC.serverName,
                world = world.name,
                x = loc.x,
                y = loc.y,
                z = loc.z,
                yaw = loc.yaw,
                pitch = loc.pitch,
            )
        }
    }

    fun toLocation(): Location? {
        val resolvedWorld = Bukkit.getWorld(world ?: return null) ?: return null
        return Location(resolvedWorld, x, y, z, yaw, pitch)
    }

    fun distance(location: Location): Double? {
        if (!isSameServer()) return null
        val locationWorld = location.world ?: return null
        if (world != locationWorld.name) return null
        val dx = location.x - x
        val dy = location.y - y
        val dz = location.z - z
        return sqrt(dx * dx + dy * dy + dz * dz)
    }

    fun isSameServer(): Boolean =
        server?.equals(ARC.serverName, ignoreCase = true) == true
}
